src/vs/workbench/contrib/preferences/browser/settingsTree.ts

```diff
@@ -338,27 +339,60 @@ export function resolveConfiguredUntrustedSettings(groups: ISettingsGroup[], tar
 	return [...allSettings].filter(setting => setting.restricted && inspectSetting(setting.key, target, configurationService).isConfigured);
 }
 
-export function resolveExtensionsSettings(groups: ISettingsGroup[]): ITOCEntry<ISetting> {
-	const settingsGroupToEntry = (group: ISettingsGroup) => {
+export async function resolveExtensionsSettings(extensionService: IExtensionService, groups: ISettingsGroup[]): Promise<ITOCEntry<ISetting>> {
+	const extGroupTree = new Map<string, ITOCEntry<ISetting>>();
+	const addEntryToTree = (extensionId: string, extensionName: string, childEntry: ITOCEntry<ISetting>) => {
+		if (!extGroupTree.has(extensionId)) {
+			const rootEntry = {
+				id: extensionId,
+				label: extensionName,
+				children: [childEntry]
+			};
+			extGroupTree.set(extensionId, rootEntry);
+		} else {
+			extGroupTree.get(extensionId)!.children!.push(childEntry);
+		}
+	};
+	const processGroupEntry = async (group: ISettingsGroup) => {
 		const flatSettings = arrays.flatten(
 			group.sections.map(section => section.settings));
 
-		return {
+		const extensionId = group.extensionInfo!.id;
+		const extension = await extensionService.getExtension(extensionId);
+		let extensionName = group.title;
+		if (extension) {
+			extensionName = extension.displayName ?? extension.name;
+		}
+
+		const childEntry = {
 			id: group.id,
 			label: group.title,
 			settings: flatSettings
-		};
+		} as ITOCEntry<ISetting>;
+		addEntryToTree(extensionId, extensionName, childEntry);
 	};
 
-	const extGroups = groups
+	const processPromises = groups
 		.sort((a, b) => a.title.localeCompare(b.title))
-		.map(g => settingsGroupToEntry(g));
+		.map(g => processGroupEntry(g));
+
+	return Promise.all(processPromises).then(() => {
+		const extGroups: ITOCEntry<ISetting>[] = [];
+		for (const value of extGroupTree.values()) {
+			if (value.children!.length === 1) {
+				// only one child,so no need to nest it
+				extGroups.push(value.children![0]);
```

Comment:
This is interesting, on one hand we may not need a bunch of single-child nodes, but on the other, now some of the nodes under "Extensions" are the extension name, but some others are the section title, which feels inconsistent. I still won't be able to tell which extension some config sections come from